File: backend/apps/accounts/chats.py
"""Private 1:1 chats."""

from django.db.models import Q
from rest_framework.exceptions import NotFound, ValidationError

from .models import DirectMessage, User

MAX_BODY = 2000


def other_party(row, me):
    return row.recipient if row.sender_id == me.id else row.sender


def require_other(me, raw_id):
    try:
        pk = int(raw_id)
    except (TypeError, ValueError):
        raise NotFound("User not found.") from None
    if pk == me.id:
        raise ValidationError({"detail": "You cannot message yourself."})
    other = User.objects.filter(pk=pk).first()
    if other is None:
        raise NotFound("User not found.")
    return other


def serialize_message(row, me):
    created = row.created_at.isoformat()
    if created.endswith("+00:00"):
        created = created[:-6] + "Z"
    return {
        "id": row.id,
        "sender_id": row.sender_id,
        "recipient_id": row.recipient_id,
        "body": row.body,
        "created_at": created,
        "mine": row.sender_id == me.id,
    }


def list_chats(me):
    rows = (
        DirectMessage.objects.filter(Q(sender=me) | Q(recipient=me))
        .select_related("sender", "recipient")
        .order_by("-id")
    )
    seen = set()
    out = []
    for row in rows:
        other = other_party(row, me)
        if other.id in seen:
            continue
        seen.add(other.id)
        out.append((other, row))
    return out


def thread_messages(me, other):
    return DirectMessage.objects.filter(
        Q(sender=me, recipient=other) | Q(sender=other, recipient=me)
    ).order_by("id")[:200]


def send_message(me, raw_id, body):
    other = require_other(me, raw_id)
    text = (body or "").strip()
    if not text:
        raise ValidationError({"body": ["Message cannot be empty."]})
    if len(text) > MAX_BODY:
        raise ValidationError({"body": ["Message is too long."]})
    return DirectMessage.objects.create(sender=me, recipient=other, body=text)

File: api/chats.js
/** Private 1:1 chats (story 12 — private messages). */
const { publicPreview } = require("./friends");

const MAX_BODY = 2000;

function httpError(status, detail) {
  const err = new Error(detail);
  err.status = status;
  return err;
}

function userRow(db, id) {
  return db.prepare("SELECT * FROM users WHERE id=?").get(id);
}

function serializeMessage(row, meId) {
  return {
    id: row.id,
    sender_id: row.sender_id,
    recipient_id: row.recipient_id,
    body: row.body,
    created_at: row.created_at,
    mine: row.sender_id === meId,
  };
}

function otherId(row, meId) {
  return row.sender_id === meId ? row.recipient_id : row.sender_id;
}

function requireOther(db, meId, rawId) {
  const id = Number(rawId);
  if (!Number.isFinite(id) || id < 1) throw httpError(404, "User not found.");
  if (id === meId) throw httpError(400, "You cannot message yourself.");
  const other = userRow(db, id);
  if (!other) throw httpError(404, "User not found.");
  return other;
}

function listChats(db, meId) {
  const rows = db
    .prepare(
      `SELECT * FROM direct_messages
       WHERE sender_id=? OR recipient_id=?
       ORDER BY id DESC`,
    )
    .all(meId, meId);
  const seen = new Set();
  const out = [];
  for (const row of rows) {
    const oid = otherId(row, meId);
    if (seen.has(oid)) continue;
    seen.add(oid);
    const other = userRow(db, oid);
    if (!other) continue;
    out.push({
      user: publicPreview(db, other, meId),
      last_message: serializeMessage(row, meId),
    });
  }
  return out;
}

function getThread(db, meId, rawId) {
  const other = requireOther(db, meId, rawId);
  const messages = db
    .prepare(
      `SELECT * FROM direct_messages
       WHERE (sender_id=? AND recipient_id=?) OR (sender_id=? AND recipient_id=?)
       ORDER BY id ASC
       LIMIT 200`,
    )
    .all(meId, other.id, other.id, meId)
    .map((row) => serializeMessage(row, meId));
  return { user: publicPreview(db, other, meId), messages };
}

function sendMessage(db, meId, rawId, body) {
  const other = requireOther(db, meId, rawId);
  const text = String(body || "").trim();
  if (!text) throw httpError(400, "Message cannot be empty.");
  if (text.length > MAX_BODY) throw httpError(400, "Message is too long.");
  const info = db
    .prepare(
      `INSERT INTO direct_messages (sender_id, recipient_id, body, created_at)
       VALUES (?, ?, ?, ?)`,
    )
    .run(meId, other.id, text, new Date().toISOString());
  const row = db.prepare("SELECT * FROM direct_messages WHERE id=?").get(info.lastInsertRowid);
  return serializeMessage(row, meId);
}

module.exports = { listChats, getThread, sendMessage };
